server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -57,12 +60,59 @@ public CoordinatorStats run(DruidCoordinator coordinator, DruidCoordinatorRuntim
 
     final Map<String, Integer> loadStatus = Maps.newHashMap();
 
+    final Map<String, Integer> tieredReplicants = getTieredReplicants();
+    for (final String tier : tieredReplicants.keySet()) {
+      stats.addToTieredStat(ASSIGNED_COUNT, tier, 0);
+    }
+
+    final BalancerStrategy strategy = params.getBalancerStrategy();
+
+    final int maxSegmentsInNodeLoadingQueue = params.getCoordinatorDynamicConfig()
+                                                    .getMaxSegmentsInNodeLoadingQueue();
+
+    final Predicate<ServerHolder> serverHolderPredicate;
+    if (maxSegmentsInNodeLoadingQueue > 0) {
+      serverHolderPredicate = s -> (s != null && s.getNumberOfSegmentsInQueue() < maxSegmentsInNodeLoadingQueue);
+    } else {
+      serverHolderPredicate = Objects::nonNull;
+    }
+
     int totalReplicantsInCluster = params.getSegmentReplicantLookup().getTotalReplicants(segment.getIdentifier());
-    for (Map.Entry<String, Integer> entry : getTieredReplicants().entrySet()) {
+
+    final Optional<ServerHolder> primaryHolderToLoad;
+    if (totalReplicantsInCluster <= 0) {
+      log.trace("No replicants for %s", segment.getIdentifier());
+      primaryHolderToLoad = getPrimaryHolder(
+          params.getDruidCluster(),
+          tieredReplicants,
+          strategy,
+          segment,
+          serverHolderPredicate
+      );
+
+      if (primaryHolderToLoad.isPresent()) {
+        final ServerHolder holder = primaryHolderToLoad.get();
+
+        holder.getPeon().loadSegment(segment, null);
+        stats.addToTieredStat(ASSIGNED_COUNT, holder.getServer().getTier(), 1);
+        ++totalReplicantsInCluster;
+      } else {
+        log.trace("No primary holder found for %s", segment.getIdentifier());
+        return stats;
```

Comment:
this is hidden in a weird spot. Is this the same behavior as previously?

Reply:
Also suggest moving this up to the top of the if statement and flipping the boolean for ease of readability.

Reply:
Sure, I'll move it up.

And yeah, I think I have mistakenly skipped the "drop" part; I will add that in.